### PR TITLE
integration tests: test sub (Pub/Sub trigger)

### PR DESCRIPTION
**Pull request by listx**
Please see the individual commits for a rundown of how things work. There are 3 parts to this PR:

- Deployment of the existing `sub` component into the test cluster
- Creation/deployment of the new `fakepubsub` service into the test cluster
- Addition of the new `TestPubSubSubscriptions` test, which currently only tests that a statically-defined postsubmit job can be triggered by publishing a Pub/Sub message (to be processed by `sub`)

/assign @cjwagner 
/cc @chaodaiG @mpherman2

### CODE REVIEW

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (chaodaiG) [inline prow/test/integration/internal/fakepubsub/fakepubsub.go:44]: would like to point out that this sets up a global env throughout the entire test process, which might or might not cause side effect. It would be ideal if this could be an option to `pubsub.NewClient`, but reading through https://cloud.google.com/pubsub/docs/emulator and there is not clear indicator of such an option.

One thing not super clear from the emulator doc, is that at which step this env var is needed, if it's not required for creating the client, aka the emulator endpoint value is assigned to a field in the client, then it might be safe to unset the env var right after the `pubsub.NewClient` call

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (chaodaiG) [inline prow/test/integration/internal/fakepubsub/fakepubsub.go:58]: might want to pass in a context, so that you can do things like inherit timeout etc.

**Review by chaodaiG — COMMENTED** [review]
Looks great, left a few comments, mostly styling I think

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (listx) [reply prow/test/integration/cmd/fakepubsub/main.go:92]: I initially agreed with your comment, but then when refactoring `initEmulatorState` to extract the config.Agent, didn't like how now the error handling for extracting the configAgent would be done in the main function, cluttering it up. This hurts my own readability and I'd prefer to keep `main` small.

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (listx) [reply prow/test/integration/cmd/fakepubsub/main.go:136]: See my comment about startPubSubEmulator()

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (listx) [reply prow/test/integration/cmd/fakepubsub/main.go:146]: Unsetting feels unsafe because we would lose (or toy with, for no benefit) the guarantee that the fakepubsub binary will only ever talk to the emulator running inside the same container, not the real Pub/Sub.

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (chaodaiG) [reply prow/test/integration/cmd/fakepubsub/main.go:92]: yeah, although majority of prow components extracts config.Agent in `main` function, it's not a good enough reason to ask you to do so.

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (listx) [reply prow/test/integration/cmd/fakepubsub/main.go:142]: > my initial reaction to initialize pubsub topics is that it should be done by client(tests) instead of server

I agree with your taste. But the additional complexity that is hidden here is sub's own startup code to set up its PullServer. If we don't  run the topic/subscription creation ASAP (which is what we're doing here), `sub` will crashloop because fakepubsub will tell sub that the subscription it wants to subscribe to does not exist. So if we move this to the test case, sub will not be in a "ready" state (which we wait for in setup-prow-components.sh) and we will fail the integration test setup phase before we even get to the tests.

File: prow/test/integration/config/prow/cluster/fakepubsub.yaml
Comment (listx) [reply prow/test/integration/config/prow/cluster/fakepubsub.yaml:5]: I really want to stay consistent with the existing fake services named fake*. If we do want to change it to `pubsub-emulator` we should also rename fakegitserver to just `gitserver` because it's not even an emulator (it's the real deal).

But yeah it's so much nicer to see "fake*" in the codebase when writing tests and configuring things because of the consistency IMHO.

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (listx) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go:58]: Oh actually I think I can just delete this (we don't use it).

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (listx) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go:44]: I don't see anything either on https://pkg.go.dev/cloud.google.com/go/pubsub#hdr-Emulator. From reading those docs I think you have to have the environment variable set at all times.

I think it's fine for now to have all pubsub client calls go to the fakepubsub service for all of our integration tests though. If we do need to talk to the real Pub/Sub service, we can revisit this.

File: prow/test/integration/config/prow/cluster/sub.yaml
Comment (listx) [reply prow/test/integration/config/prow/cluster/sub.yaml]: Done

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (listx) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go:51]: Oops, I do see a field being set actually in https://github.com/googleapis/google-cloud-go/blob/e43c095c94e44a95c618861f9da8f2469b53be16/pubsub/pubsub.go#L126. Will update this.

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (chaodaiG) [inline prow/test/integration/internal/fakepubsub/fakepubsub.go:77]: same as last comment, might want to pass in a context

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (chaodaiG) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go:51]: nice

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (listx) [reply prow/test/integration/cmd/fakepubsub/main.go:146]: We can use the same trick of setting the client address field for this that I found in the other comment, instead of using an env var here. For the containerized binary it doesn't matter but it does matter if we ever want to debug locally. Will update this.

Comment (chaodaiG) [reply prow/test/integration/cmd/fakepubsub/main.go:146]: yeah I would much prefer that instead of env var, great finding!

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (listx) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go]: Done

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (listx) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go]: Deleted

File: prow/test/integration/internal/fakepubsub/fakepubsub.go
Comment (listx) [reply prow/test/integration/internal/fakepubsub/fakepubsub.go]: Done

File: prow/test/integration/cmd/fakepubsub/main.go
Comment (listx) [reply prow/test/integration/cmd/fakepubsub/main.go]: Done

File: .ko.yaml
Comment (chaodaiG) [inline .ko.yaml:50]: please ping to a version instead of latest to ensure deterministic outcome

Comment (listx) [reply .ko.yaml]: Done

**Review by chaodaiG — COMMENTED** [review]
/lgtm
/approve

still need to fix tests though

**Review by chaodaiG — APPROVED** [review]
/hold
this PR changes global go.mod, better not merge right before weekend

**Pull request merged**
